financial-templates-lib/ExpiringMultiPartyEventClient.js

```diff
@@ -0,0 +1,165 @@
+class ExpiringMultiPartyEventClient {
+  constructor(abi, web3, empAddress, updateThreshold = 60) {
+    this.updateThreshold = updateThreshold;
+    this.lastUpdateTimestamp;
+
+    this.web3 = web3;
+
+    // EMP contract
+    this.emp = new web3.eth.Contract(abi, empAddress);
+    this.empAddress = empAddress;
+
+    // EMP Events
+    this.liquidationEvents = [];
+    this.disputeEvents = [];
+    this.disputeSettlementEvents = [];
+
+    this.highestBlockNumberSeen = 0;
+  }
+
+  // Calls _update unless it was recently called, as determined by this.updateThreshold.
+  update = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    if (currentTime < this.lastUpdateTimestamp + this.updateThreshold) {
+      Logger.debug({
+        at: "ExpiringMultiPartyEventClient",
+        message: "EMP state update skipped",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTimestamp,
+        timeRemainingUntilUpdate: this.lastUpdateTimestamp + this.updateThreshold - currentTime
+      });
+      return;
+    } else {
+      await this._update();
+      this.lastUpdateTimestamp = currentTime;
+      Logger.debug({
+        at: "ExpiringMultiPartyEventClient",
+        message: "EMP state updated",
+        lastUpdateTimestamp: this.lastUpdateTimestamp
+      });
+    }
+  };
+
+  // Force call of _update, designed to be called by downstream caller that knowingly updated the EMP state.
+  forceUpdate = async () => {
+    const currentTime = Math.floor(Date.now() / 1000);
+    await this._update();
+    this.lastUpdateTimestamp = currentTime;
+    Logger.debug({
+      at: "ExpiringMultiPartyEventClient",
+      message: "EMP state force updated",
+      lastUpdateTimestamp: this.lastUpdateTimestamp
+    });
+  };
+
+  // Delete all events within the client
+  clearState = async () => {
+    this.liquidationEvents = [];
+    this.disputeEvents = [];
+    this.disputeSettlementEvents = [];
+  };
+
+  // Returns an array of liquidation events.
+  getAllLiquidationEvents = () => this.liquidationEvents;
+
+  // Returns an array of dispute events.
+  getAllDisputeEvents = () => this.disputeEvents;
+
+  // Returns an array of dispute events.
+  getAllDisputeSettlementEvents = () => this.disputeSettlementEvents;
+
+  start = () => {
+    this._poll();
+  };
+
+  _poll = async () => {
+    while (true) {
+      try {
+        await this._update();
+      } catch (error) {
+        Logger.error({
+          at: "ExpiringMultiPartyEventClient",
+          message: "client polling error",
+          error: error
+        });
+      }
+      await delay(Number(10_000));
+    }
+  };
+
+  _update = async () => {
+    // If there have been previous events retrieve we should only query the chain from
+    // the last seen block number. This max block checks for the highest block number
+    // seen across the different internal data structures.
+    const highestBlockNumberSeen = Math.max(
```

Comment:
I think this definitely works, and it's a totally reasonable implementation.

However, after rereading my comment, I realized I didn't communicate the suggestion very well. What I was suggesting was something like the following:
```
const currentBlockNumber = await web3.eth.getBlockNumber();

const liquidationEventsObj = await this.emp.getPastEvents("LiquidationCreated", {
    fromBlock: this.lastBlockNumberSeen,
    toBlock: currentBlockNumber
});
...
...
...
this.lastBlockNumberSeen = currentBlockNumber;
```
The only reason I suggest this is that even if there are no events in the blocks, you're still skipping them if you've already queried them. I think it's also a *little* easier to read. Also, if you want to control for number of confirmations before you "accept" a block, you can easily do so by doing `(await web3.eth.getBlockNumber()) - numConfirmations` so you only query up to `numConfirmations` blocks ago.

Totally up to you which one you wanna go with. Just wanted to re-communicate so the suggestion was clearer.

Reply:
ok, great. I agree this is better, much cleaner like this in your suggestion. I was avoiding the async web3 call but the readability is better in your way. have updated accordingly.